upstream/upstream.go

```diff
@@ -654,11 +666,91 @@ func (u *Upstream) shouldSkip(req *common.NormalizedRequest) (reason error, skip
 		}
 	}
 
-	// TODO evm: if block can be determined from request and upstream is only full-node and block is historical skip
+	// if block can be determined from request and upstream is only full-node and block is historical skip
+	bn, ebn := req.EvmBlockNumber()
+	if ebn != nil || bn <= 0 {
+		return fmt.Errorf("failed to get valid request block number: %w", ebn), false
+	}
+
+	nodeType := u.config.Evm.NodeType
+	if nodeType == "" {
+		detectedNodeType, err := u.detectNodeType()
+		if err != nil {
+			return fmt.Errorf("failed to detect node type: %w", err), false
+		}
+		nodeType = common.EvmNodeType(detectedNodeType)
+	}
+
+	if nodeType == common.EvmNodeTypeFull {
+		lb := req.Network().EvmStatePollerOf(u.Config().Id).LatestBlock()
+
+		if bn < lb-u.config.Evm.MaxAvailableRecentBlocks {
+			return fmt.Errorf("block number %d is out of range (must be >= %d)", bn, lb-u.config.Evm.MaxAvailableRecentBlocks), true
```

Comment:
I think we can return common.MissingData... error here, so it's properly handled downstream